lib/enumerator.h

```diff
@@ -630,6 +631,11 @@ const EnumeratorHandle<T> &EnumeratorHandle<T>::operator++() {
     return *this;
 }
 
+template <typename T>
+bool EnumeratorHandle<T>::operator==(const EnumeratorHandle<T> &other) const {
+    return !(*this != other);
```

Comment:
Sorry for the noise comment, but is this way of defining `operator==` common in C++?

Reply:
Usually the opposite (`operator!=` is implemented from `operator==`), however, here we already having `operator!=` with quite peculiar semantics.

Reply:
As @asl pointed out in https://github.com/p4lang/p4c/issues/5107#issuecomment-2608881119, it's probably better to first define `operator==` and then define `operator!=` as `!(*this == other)`, which is also the semantics specified in https://en.cppreference.com/w/cpp/named_req/InputIterator. I still don't have a good understanding here to make that further simplification. Maybe someone with a better understanding can do it in a separate PR.

Reply:
> Usually the opposite (`operator!=` is implemented from `operator==`),

Indeed, with C++20, you don't even need to do that -- if you define `operator==`, the compiler will synthesize `operator!=` for you